src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,267 @@ PhaseStatus Compiler::optUnrollLoops()
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("After if conversion\n");
+        for (BasicBlock* dump_block = block; dump_block != middle_block->bbNext; dump_block = dump_block->bbNext)
+        {
+            fgDumpBlock(dump_block);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    return true;
+#endif
+}
+
+bool Compiler::optIfConvertSelect(GenTree* original_condition, GenTree* asg_node)
+{
+    assert(original_condition->OperIsCmpCompare());
+    assert(asg_node->gtOper == GT_ASG);
+
+    // Duplicate the input of the assign.
+    // This will be used as the false result of the select node.
+    GenTree* current_value = gtCloneExpr(asg_node->AsOp()->gtOp1);
+    current_value->gtFlags &= GTF_EMPTY;
+
+    // Duplicate the condition and invert it
+    GenTree* cond = gtCloneExpr(original_condition);
+    cond->gtFlags |= GTF_DONT_CSE;
+    cond->gtFlags ^= GTF_RELOP_JMP_USED;
+    cond->gtFlags ^= GTF_RELOP_NAN_UN;
+    cond->gtOper = GenTree::ReverseRelop(cond->gtOper);
+
+    // Create a select node.
+    GenTreeConditional* select =
+        gtNewConditionalNode(GT_SELECT, cond, asg_node->AsOp()->gtOp2, current_value, asg_node->TypeGet());
+
+    // Use the select as the input to the assignment.
+    asg_node->AsOp()->gtOp2 = select;
+    asg_node->AsOp()->gtFlags |= (select->gtFlags & GTF_ALL_EFFECT);
+
+    // Calculate costs.
+    gtSetEvalOrder(select);
+
+    return true;
+}
+
+bool Compiler::optIfConvertCCmp(GenTree* original_condition, GenTree* asg_node)
+{
+    assert(original_condition->OperIsCmpCompare());
+    assert(asg_node->gtOper == GT_ASG);
+
+    // For now, don't handle floats.
+    if (!varTypeIsIntegralOrI(original_condition->AsOp()->gtOp1->TypeGet()))
+    {
+        return false;
+    }
+
+    // Note: Limiting the maximum number of chained compares may give a performance
+    //       boost, at the cost of more emitted code.
+
+    GenTreeConditional* select_node      = asg_node->AsOp()->gtOp2->AsConditional();
+    GenTree*            select_node_cond = select_node->gtCond;
+
+    // Duplicate the condition and invert it
+    GenTree* new_cond = gtCloneExpr(original_condition);
+    new_cond->gtFlags |= GTF_DONT_CSE;
+    new_cond->gtFlags ^= GTF_RELOP_JMP_USED;
+    new_cond->gtFlags ^= GTF_RELOP_NAN_UN;
+    new_cond->gtOper = GenTree::ReverseRelop(new_cond->gtOper);
+
+    // Link together the two conditions.
+    GenTree* const link = gtNewOperNode(GT_AND, TYP_INT, select_node_cond, new_cond);
+    link->gtFlags |= (select_node_cond->gtFlags & GTF_ALL_EFFECT);
+    link->gtFlags |= (new_cond->gtFlags & GTF_ALL_EFFECT);
+
+    // Use the link as the conditional input to the select.
+    select_node->gtCond = link;
+    select_node->gtFlags |= (link->gtFlags & GTF_ALL_EFFECT);
+    select_node->gtFlags ^= GTF_RELOP_NAN_UN;
+    asg_node->AsOp()->gtFlags |= (select_node->gtFlags & GTF_ALL_EFFECT);
+
+    // Calculate costs.
+    gtSetEvalOrder(link);
+
+    return true;
+}
```

Comment:
Can you give an example of how the IR nodes are changing when this case triggers?
It's likely this is not the right place for this optimization. It should probably go in `optOptimizeBools` instead.

Reply:
Additionally, it will be good to represent an example tree in the comments (see function headers for `optOptimizeBools`)  for easy visualization.

Reply:
>Can you give an example of how the IR nodes are changing when this case triggers?

Added a big block of comments, plus renamed the helper functions.

>It's likely this is not the right place for this optimization. It should probably go in optOptimizeBools instead.

Hmm. I'm assuming you mean the entire If convert phase? (as the use of ccmps relies on the select already existing).

That does mean I would have to add code for CSE and const propagation (which went down a rabbit hold of stuff when I originally tried it).

Reply:
That could work....

In the AND pass, it would repeatedly look for two `JTRUE` blocks next to each other with appropriate destinations. When found they would be merged into a single `JTRUE` with its child being an `AND`.

In the If convert phase, we then need to only check for a single JTRUE block.

I think that might then make it easier to support `else` cases or multiple statements in the body. 

Sigh 😄  Ok .... in that case I'd probably want to drop the `AND` chains from this patch and just focus on the `SELECT`.

Reply:
That sounds ok to me, it should also make it easier to get this in (these transformations are going to need some profitability heuristics, I think).